DistroLauncher/ChildProcess.h

```diff
@@ -0,0 +1,165 @@
+ *  ChildProcess child{L"notepad.exe",L"",nullptr,nullptr,nullptr}; //nullptr handles disable console interaction.
+ *  ```
+ *
+ *  Set a listener if and when you want to be notified about an extraneous process termination:
+ *
+ *  ```
+ *  child.setListener([]{std::cout << "Who killed my child?\n";});
+ *  ```
+ *
+ *  > The function object above will be owned by this class. The listener will be called on another thread, be careful
+ * with synchronization. Throwing exceptions there is also not a very good idea.
+ *  > Setting a listener for a process that never runs effectively does nothing.
+ *
+ *  Kill it if you're upset with its behavior for some reason:
+ *
+ * ```
+ *  child.terminate();
+ * ```
+ *
+ * Or just let the destructor do it. Move the child to an outer scope if you don't want the process to be killed.
+ *
+ * Or else wait on the process to exit:
+ *
+ * ```
+ * DWORD exitCode = child.waitExitSync(); // without any arguments it will wait forever.
+ * ```
+ *
+ */
+
+namespace Oobe
+{
+
+    template <typename ProcessAPI> class ChildProcessInterface
+    {
+      public:
+        // Alias to just enforce the communication that the supplied callback will run on another thread.
+        using CallInOtherThread = std::function<void()>;
+
+        // Starts the process.
+        bool start()
+        {
+            return static_cast<ProcessAPI*>(this)->do_start();
+        }
+
+        /// Sets the listener that will be called back once (and if) the process ends from the outside.
+        /// IMPORTANT NOTE ON MULTITHREADING: [onClose] will be invoked by another thread, thus any precaution
+        /// required to ensure safe concurrent call must be taken by the [onClose] callable implementation itself and
+        /// any functions it calls inside its body. That's the reason for the horrible type name.
+        template <typename CallableInOtherThread> void setListener(CallableInOtherThread&& onCloseCallback)
+        {
+            onCloseListener.emplace(std::forward<CallableInOtherThread>(onCloseCallback));
+        }
+
+        // Terminates the running process forcebly.
+        void terminate()
+        {
+            unsubscribe();
+            static_cast<ProcessAPI*>(this)->do_terminate();
+        }
+        DWORD waitExitSync(DWORD timeoutMs = INFINITE)
+        {
+            return static_cast<ProcessAPI*>(this)->do_waitExitSync(timeoutMs);
+        }
+
+        // Unregisters the wait callback, if set.
+        void unsubscribe()
+        {
+            if (waiterHandle != nullptr) {
+                static_cast<ProcessAPI*>(this)->do_unsubscribe();
+                waiterHandle = nullptr;
+            }
+        }
+
+        DWORD pid() const
+        {
+            return procInfo.dwProcessId;
+        }
+
+        DWORD threadId() const
+        {
+            return procInfo.dwProcessId;
+        }
+
+        constexpr ChildProcessInterface(const std::filesystem::path& exePath, const std::wstring& args,
```

Comment:
No `std::basic_string` constructor in c++17 is constepxr, so this function can never be called from a constexpr context.

The same is true for `std::filesystem::path`.

Reply:
Indeed.